### PR TITLE
Gulp tasks test

### PR DESCRIPTION
**Pull request by DerekAhn**
## Flexibility for test configs

Users may either pass in a path to their local babelrc, file, or pass in a custom object.
Refactored for all options to be encapsulated on the test property:

```js
test: {
  paths: [strings],
  globPath: string,
  pathToBabelRc: string,
  customBabelConfig: object
}
```

* If both a pathToBabelRc and a customBabelConfig, the customBabelConfig will trump.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -89,9 +89,13 @@ var opts = {
...
   distPreprocess: ['generate-index-icons', 'dist-css'],
   scsslint: true,
-  testPaths: [
-    'test/**/*.js'
-  ]
+
+  test: {
+    paths: ['test/**/*.js'],
+    // You may pass in a path or a custom config obj
+    pathToBabelRc: __dirname+'/.babelrc'
```
Comment (alansouzati) [inline gulpfile.js]: I'm wondering if there is a way to automatically identify this without exposing the path to babel.

If you move this to be inside ` src/utils/gulp/gulp-tasks-test.js` would that work?

Ultimately, I believe we could achieve this:
  1 - Try to load `require('fs').readFileSync(path.join(__dirname, '.babelrc'), 'utf8')`
  2 - If root .babelrc is not loaded, use the default one.

File: gulpfile.js [hunk 2]
```diff
@@ -89,9 +89,13 @@ var opts = {
...
+    // customBabelConfig: Object.assign({}, custom babelrc object)
```
Comment (alansouzati) [inline gulpfile.js]: can you move `customBabelConfig` to be in the root level of our options? 

Babel is something we are using elsewhere, so it is not specific to the test context.

File: src/utils/gulp/gulp-tasks-test.js [hunk 3]
```diff
@@ -17,38 +21,47 @@ module.exports = function(gulp, options) {
   };
 
   gulp.task('test', function(done) {
-    if (options.testPaths) {
```
Comment (alansouzati) [inline src/utils/gulp/gulp-tasks-test.js]: If we decide to move away from `testPaths` we need to add backward compatibility, meaning we should deprecated the options while still supporting it for now.

But, I believe we can maintain `options.testPaths` as is.

File: src/utils/gulp/gulp-tasks-test.js [hunk 4]
```diff
@@ -17,38 +21,47 @@ module.exports = function(gulp, options) {
...
       var argv = require('yargs').argv;
 
+      // customBabelRc object, or path to local babelrc
+      var customBabelConfig = test.customBabelConfig?
+        test.customBabelConfig :
+        typeof test.pathToBabelRc === 'string'?
+        JSON.parse(require('fs').readFileSync(test.pathToBabelRc, 'utf8')) :
+        {};
+
+      var babelConfig = merge({}, {
```
Comment (alansouzati) [inline src/utils/gulp/gulp-tasks-test.js]: Awesome, I like you are using merge here.

File: src/utils/gulp/gulp-tasks-test.js [hunk 5]
```diff
@@ -17,38 +21,47 @@ module.exports = function(gulp, options) {
...
+        "plugins": [
+          "transform-object-rest-spread",
+          "add-module-exports"
+        ]
+      }, customBabelConfig);
+
       jsxCoverage.initModuleLoaderHack(jsxCoverageOptions);
 
-      glob.sync('**src/js/**/*.js').forEach(function (file) {
+      glob.sync(test.globPath || '**src/js/**/*.js').forEach(function (file) {
```
Comment (alansouzati) [inline src/utils/gulp/gulp-tasks-test.js]: sorry i add that string here. But I believe we can safely replace that like with `options.jsAssets`

**Comment by alansouzati** [comment]
Thanks for your contribution I believe we can maintain the structure as it is `options.testPaths`. I like your proposal of adding a  `customBabelConfig` prop, but I believe it should not be restricted to the test environment. 

I don't think we need `globPath` and `pathToBabelRc`

File: src/utils/gulp/gulp-tasks-test.js (hunk 5 above)
Comment (DerekAhn) [inline src/utils/gulp/gulp-tasks-test.js]: Do you mean to pass the whole array?

File: gulpfile.js (hunk 2 above)
Comment (DerekAhn) [inline gulpfile.js]: Should a customConfig be a toggle then? Like a property "useLocalBabelRc: bool"?

**Pull request closed without merging**